src/libraries/System.Text.Json/docs/JsonSerializationSourceGeneration.md

```diff
@@ -0,0 +1,238 @@
+
+#### Implicit Model
+Various implicit approaches have been discussed such as source generating all partial classes or scanning for calls into the serializer using the Roslyn tree scanning. These models can be revisited in the future as the value/feasibility of the approach becomes clearer based on user feedback. It is important to note that some downsides to such model can result in missing types to source generate or source generating types when not needed due to a bug or edge cases we didn’t consider. 
+
+#### Proposed Explicit Model
+There are two scenarios within the proposed explicit model:
+
+1. Base case: Code generates a facade ```JsonClassInfo``` for the attribute target class/struct.
+
+```c#
+ // (Base Case) Codegen (De)Serialization ExampleLoginClassInfo. 
+[JsonSerializable] 
+public class ExampleLogin
+{ 
+    public string Email { get; set; } 
+    public string Password { get; set; } 
+}
+```
+2. Pass in type: Code generates a ```JsonClassInfo``` for the attribute target class/struct using the passed in type. This scenario can be used if you don't own the serializable class.
+
+```c#
+// (Pass in Type) Codegen (De)Serialization to create ExampleExternalClassInfo using ExternClass. 
+[JsonSerializable(typeof(ExternalClass))] 
+public static class ExampleExternal { }
+```
+
+The output of this phase would be a list of reflection-type-like models where we can iterate through the type's members in order to codegen recursively. The scope of this phase is to only find the root serializable types instead of the whole type-graph since we want to recursively codegen without storing the whole type-graph in-memory. 
+
+We believe that an explicit model using attributes would be a simple first-approach to the problem given that the source code generation needs the user to declare their type class as partial anyway. We can then use Roslyn tree to find the JsonSerializable attribute for both types the user owns and doesn’t own to source generate using Roslyn Source Generators.
+
+#### New API Proposal
+
+```C#
+namespace System.Text.Json.Serialization
+{
+    /// <summary>
+    /// When placed on a type, will source generate (de)serialization for the specified type and all types in it's object graph.
+    /// </summary>
+    /// <remarks>
+    /// Must take into account that type discovery using this attribute is at compile time using Source Generators.
+    /// </remarks>
+    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Struct, AllowMultiple = false)]
+    public sealed class JsonSerializableAttribute : JsonAttribute
+    {
+        /// <summary>
+        /// Takes target class/struct to construct a facade JsonClassInfo as TargetNameClassInfo.
+        /// </summary>
+        public JsonSerializableAttribute() { }
+
+        /// <summary>
+        /// Takes type as an argument and uses it to create a facade JsonClassInfo as TargetNameClassInfo.
+        /// </summary>
+        public JsonSerializableAttribute(Type type) { }
```

Comment:
I would like a copy of the ctor here that would optionally take in ``JsonSerializerOptions`` to use with the source generator so that the generated source would use those specific options instead of possibly the default (in cases where wanting to (de)serialize the json might fail or have issues without those options explicitly being set.

Reply:
I believe https://github.com/dotnet/runtimelab/pull/16#discussion_r577094141 addresses the concern here - https://github.com/dotnet/runtimelab/pull/16#discussion_r577094141.